design/kata-api-design.md

```diff
@@ -0,0 +1,143 @@
+-  sandbox.Resume()
+
+Resume the paused sandbox.
+-  sandbox.Destroy()
+
+Destroy the sandbox and remove all persistent metadata.
+-  sandbox.Stats()
+
+Get the stats of the sandbox.
+-  sandbox.Done()
+
+Return a context handler that caller can monitor for sandbox callbacks such as
+error termination.
+-  sandbox.ListContainer()
+
+List all containers in the sandbox.
+-  sandbox.CreateContainer()
+
+Create new container in the sandbox.
+-  sandbox.DeleteContainer()
+
+Delete a container from the sandbox.
+-  sandbox.StartContainer()
+
+Start a container in the sandbox.
+-  sandbox.StatusContainer()
+
+Get the status of a container in the sandbox.
+-  sandbox.AddProcess()
+
+Run a new process in the sandbox.
+-  sandbox.WaitProcess()
+
+Wait on a process.
+### Sandbox Hotplug API
+-  sandbox.AddStorage()
+
+Add new storage to the sandbox.
+-  sandbox.AddNic()
+
+Add new nic to the sandbox.
+-  sandbox.UpdateRoute()
+
+### Sandbox Relay API
+Update the sandbox route table, e.g. for portmapping support.
+-  sandbox.ContainerTTYResize(containerID, processID, Height, Width)
+
+Relay TTY resize request to a process.
+-  sandbox.SignalProcess(containerID, processID, signalALL)
+
+Relay a signal to a process or all processes in a container.
+-  sandbox.ProcessStdio(containerID, processID)
+
+Relay a process stdio. Return stdin/stdout/stderr pipes to the process stdin/stdout/stderr streams.
+## Plugin framework for external proprietary Kata runtime extensions
+### Hypervisor plugin
+
+TBD.
+### Metadata storage plugin
+The metadata storage plugin controls where sandbox metadata is saved.
+   -  storage.Save(key, value)
+
+Save a record.
+   -  storage.Load(key)
+
+Load a record.
+   -  storage.Delete(key)
+
+Delete a record.
+Built-in implementations include:
+   -  Filesystem storage
+   -  LevelDB storage
+### VM Factory plugin
+The VM factory plugin controls how a sandbox factory creates new VMs.
+   -  VMFactory.NewVM(HypervisorConfig)
+
+Create a new VM based on `HypervisorConfig`.
+
+Built-in implementations include:
+   -  CreateNew()
+
+Create brand new VM based on HypervisorConfig
+   -  CreateFromTemplate()
+
+Create new VM from template.
+   -  CreateFromCache()
+
+Create new VM from VM caches.
+### Sandbox Creation Plugin Workflow
+![Sandbox Creation Plugin Workflow](https://raw.githubusercontent.com/bergwolf/raw-contents/master/kata/Kata-sandbox-creation.png "Sandbox Creation Plugin Workflow")
+
+### Sandbox Connection Plugin Workflow
+![Sandbox Connection Plugin Workflow](https://raw.githubusercontent.com/bergwolf/raw-contents/master/kata/Sandbox-Connection.png "Sandbox Connection Plugin Workflow")
+
+## Built-in Shim and Proxy Types and Capabilities
+-  Built-in shim/proxy sandbox configurations
+   -  Supported Shim configurations:
+      -  noopshim
+      -  ccshim
+      -  katashim
+   -  Supported Proxy configurations:
+      -  noopProxy
+      -  noProxy
+      -  ccProxy
+      -  kataProxy
+      -  kataNoProxy
```

Comment:
I think we need descriptions for all of these as some are not obvious (for example what is the real difference between `noopProxy` and `noProxy`?)

Reply:
`noopProxy` and `noProxy` are existing types of proxy config in virtcontainers. IIUC `noopProxy` is to spawn a proxy process that does noting and it's for testing. `noProxy` means no proxy process at all.

And `kataNoProxy` means no proxy process + yamux grpc dialer, aka, proxy functionality without proxy process.

Reply:
That's right. `noProxy` and `noopProxy` are explained here:

- https://github.com/containers/virtcontainers/blob/master/no_proxy.go#L23..L32
- https://github.com/containers/virtcontainers/blob/master/noop_proxy.go#L19..L20

 I think it would be useful to have this information in the document itself though.

Reply:
You are right. I'll add explanation to each config when updating. Thanks!